src/v/cluster/tm_stm.cc

```diff
@@ -0,0 +1,401 @@
+                    + config::shard_local_cfg().tm_sync_timeout_ms.value();
+    co_await wait(model::offset(r.value().last_offset()), deadline);
+
+    ptx = _tx_table.find(tx_id);
+    if (ptx == _tx_table.end()) {
+        co_return tm_stm::op_status::conflict;
+    }
+    if (ptx->second.etag != tx.etag) {
+        co_return tm_stm::op_status::conflict;
+    }
+    co_return tm_stm::op_status::success;
+}
+
+bool tm_stm::add_partitions(
+  kafka::transactional_id tx_id,
+  tm_etag etag,
+  std::vector<tm_transaction::tx_partition> partitions) {
+    auto ptx = _tx_table.find(tx_id);
+    if (ptx == _tx_table.end()) {
+        return false;
+    }
+    if (ptx->second.etag != etag) {
+        return false;
+    }
+    ptx->second.etag = etag.inc_mem();
+    for (auto& partition : partitions) {
+        ptx->second.partitions.push_back(partition);
+    }
+    return true;
+}
+
+void tm_stm::expire_old_txs() {
+    // TODO: expiration of old transactions
+}
+
+void tm_stm::load_snapshot(stm_snapshot_header hdr, iobuf&& tm_ss_buf) {
+    vassert(
+      hdr.version == supported_version,
+      "unsupported seq_snapshot_header version {}",
+      hdr.version);
+    iobuf_parser data_parser(std::move(tm_ss_buf));
+    auto data = reflection::adl<tm_snapshot>{}.from(data_parser);
+
+    for (auto& entry : data.transactions) {
+        _tx_table.emplace(entry.id, entry);
+        _tx_locks.emplace(entry.id, ss::make_lw_shared<mutex>());
+        _end_locks.emplace(entry.id, ss::make_lw_shared<mutex>());
+    }
+    _last_snapshot_offset = data.offset;
+    _insync_offset = data.offset;
+}
+
+stm_snapshot tm_stm::take_snapshot() {
+    tm_snapshot tm_ss;
+    tm_ss.offset = _insync_offset;
+    for (auto& entry : _tx_table) {
+        tm_ss.transactions.push_back(entry.second);
+    }
+
+    iobuf tm_ss_buf;
+    reflection::adl<tm_snapshot>{}.to(tm_ss_buf, tm_ss);
+
+    stm_snapshot_header header;
+    header.version = supported_version;
+    header.snapshot_size = tm_ss_buf.size_bytes();
+
+    stm_snapshot stm_ss;
+    stm_ss.header = header;
+    stm_ss.offset = _insync_offset;
+    stm_ss.data = std::move(tm_ss_buf);
+    return stm_ss;
+}
+
+ss::future<> tm_stm::apply(model::record_batch b) {
+    const auto& hdr = b.header();
+    _insync_offset = b.last_offset();
+
+    if (hdr.type != tm_update_batch_type) {
+        return ss::now();
+    }
+
+    vassert(
+      b.record_count() == 1,
+      "We expect single command in a batch of tm_update_batch_type");
+    auto r = b.copy_records();
+    auto& record = *r.begin();
+    auto rk = reflection::adl<uint8_t>{}.from(record.release_key());
+
+    vassert(
+      rk == tx_updated_cmd::record_key, "Unknown tm update command: {}", rk);
+    tx_updated_cmd cmd = reflection::adl<tx_updated_cmd>{}.from(
+      record.release_value());
+
+    if (!_end_locks.contains(cmd.tx.id)) {
+        _end_locks.emplace(cmd.tx.id, ss::make_lw_shared<mutex>());
+    }
```

Comment:
you can take advantage of `emplace`, `try_emplace`, and `insert` to avoid two hash table look-ups (one for contains and one for emplace) all of which will not override an entry if the key already exists. the difference between the three really boils down to optimizations for specific cases.